proxy/src/telemetry/metrics/mod.rs

```diff
@@ -597,14 +627,107 @@ impl HyperService for Serve {
                 .with_status(StatusCode::NotFound));
         }
 
-        let body = {
-            let metrics = self.metrics.lock()
-                .expect("metrics lock poisoned");
-            format!("{}", *metrics)
+        let metrics = self.metrics.lock()
+            .expect("metrics lock poisoned")
+            .deref()
+            .clone();
+
+        let is_gzip = is_gzip(&req);
+        let (writer, body) = WriteBody::new();
+        let work = future::lazy(move || {
+            if is_gzip {
+                let mut writer = GzEncoder::new(Vec::<u8>::new(), Compression::Default);
+                write!(writer, "{}", metrics)
+                    .and_then(move |()| {
+                        trace!("gzipping metrics");
+                        writer.finish()
+                    })
+            } else {
+                let mut writer = Vec::<u8>::new();
+                write!(writer, "{}", metrics).map(|_| writer)
+            }
+        })
+        .and_then(|text| tokio_io::io::write_all(writer, text))
+        .and_then(|(w, _)| tokio_io::io::shutdown(w))
+        .map(|_| ())
+        .map_err(|e| {
+            error!("error writing metrics: {:?}", e);
+        });
+
+        self.handle.spawn(work);
```

Comment:
Can you explain why it's necessary to do this work in another task? It seems like needless complexity, given that we're not really _unblocking_ anything, are we?

Reply:
I thought that having this spawn a new task would free the server task to accept a new connection while a long body is still being compressed?

Reply:
I think we don't need to serve more than one `/metrics` response concurrently. However, if this is designed to keep service discovery from being blocked on the compression then it might make sense. Do we have any decent estimate of whether the compression takes more than a few milliseconds?

Reply:
Spawning another task by itself doesn't help the server. It can juggle several different futures at once (though HTTP/1 means 1 per connection at a time). It'd only matter if compression were CPU intensive, in which case, a separate task by itself isn't enough. It'd need to compress on a separate thread. I don't know if that is needed or not, depends how much time a typical `/metrics` response takes to compress.